diff --git a/pkgs/desktops/gnome/core/pop-shell-shortcuts/default.nix b/pkgs/desktops/gnome/core/pop-shell-shortcuts/default.nix
new file mode 100644
index 0000000000000..986599ed1a989
--- /dev/null
+++ b/pkgs/desktops/gnome/core/pop-shell-shortcuts/default.nix
@@ -0,0 +1,32 @@
+{ lib, stdenv, fetchFromGitHub, rustPlatform, pkg-config, gtk3 }:
+
+stdenv.mkDerivation rec {
+  pname = "pop-shell-shortcuts";
+  version = "unstable-2021-10-02";
+
+  src = fetchFromGitHub {
+    owner = "pop-os";
+    repo = "shell-shortcuts";
+    rev = "005f76c8e59d924a7edb59a2d08030d439b7fd45";
+    sha256 = "sha256-/o6sQ/EieIbZrD7p1dwYOQ+uBsk+zf3vjnPXOcD890Q=";
+  };
+
+  cargoDeps = rustPlatform.fetchCargoTarball {
+    inherit src;
+    name = "${pname}-${version}";
+    hash = "sha256-SySh4/xjP+KnuL3x+IFQUyLgpI3hyshi3h8L5m5ZXOk=";
+  };
+
+  nativeBuildInputs = [ pkg-config rustPlatform.cargoSetupHook rustPlatform.rust.cargo ];
+  buildInputs = [ gtk3 ];
+
+  installFlags = [ "prefix=$(out)" "DESTDIR=" ];
+
+  meta = with lib; {
+    description = "Application for displaying and demoing Pop Shell shortcuts";
+    license = licenses.gpl3Only;
+    homepage = "https://github.com/pop-os/shell-shortcuts";
+    platforms = platforms.linux;
+    maintainers = with maintainers; [ mapydayv7 ];
+  };
+}
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index 0de525a34a16b..9368e29414701 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -31071,6 +31071,8 @@ with pkgs;
 
   pop-launcher = callPackage ../desktops/gnome/core/pop-launcher { };
 
+  pop-shell-shortcuts = callPackage ../desktops/gnome/core/pop-shell-shortcuts { };
+
   latte-dock = libsForQt5.callPackage ../applications/misc/latte-dock { };
 
   gnome-themes-extra = gnome.gnome-themes-extra;
