diff --git a/tools.go b/tools.go
index 6aae64c..9bd7612 100644
--- a/tools.go
+++ b/tools.go
@@ -277,6 +277,14 @@ func clientMenuContext(class string, instances []client) gtk.Menu {
 			log.Debugf("%s -> %s", cmd, reply)
 		})

+		subitem, _ = gtk.MenuItemNewWithLabel("minimize")
+		submenu.Append(subitem)
+		subitem.Connect("activate", func() {
+			cmd := fmt.Sprintf("dispatch movetoworkspacesilent special:minimized,address:%s", a)
+			reply, _ := hyprctl(cmd)
+			log.Debugf("%s -> %s", cmd, reply)
+		})
+
 		subitem, _ = gtk.MenuItemNewWithLabel("fullscreen")
 		submenu.Append(subitem)
 		subitem.Connect("activate", func() {
